Sample config and README update. (#12)
WiMoVE-OSS/wimoved: README.md

# WiMoVE - Wireless Mobility through VXLAN EVPN

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/getcursor/cursor/blob/main/LICENSE)
![CI Status](https://github.com/WiMoVE-OSS/wimoved/actions/workflows/build.yml/badge.svg?branch=main)
![Linter Status](https://github.com/WiMoVE-OSS/wimoved/actions/workflows/linter.yml/badge.svg?branch=main)

WiMoVE is a scalable Wi-Fi System that partitions all stations in separate overlay L2 domains, to limit the amount of wireless L2 broadcast traffic. In large Wi-Fi systems broadcast traffic can take up large amounts of airtime. A great talk, that explains the problem in more detail is available [here](https://www.youtube.com/watch?v=v8y-r9JBhmw). The overlay L2 domains "follow" the stations, being resized on demand, thus preserving handover.

WiMoVE is built with standard network protocols, on top of open&#8209;source technology:

- The overlay networks use [BGP EVPN](https://www.rfc-editor.org/rfc/rfc8365.html) with [VXLAN](https://www.rfc-editor.org/rfc/rfc7348) encapsulation.
- All BGP speakers run [FRRouting](https://frrouting.org/) to exchange EVPN routes.

> - The access points run [OpenWRT](https://openwrt.org/) with a custom, open&#8209;source daemon called [wimoved](https://github.com/wimove-oss/wimoved).

Reviewer: ```suggestion
- The access points run [OpenWRT](https://openwrt.org/) with our custom, open&#8209;source daemon called [wimoved](https://github.com/wimove-oss/wimoved).
```

This solution allows for using commodity access points running OpenWRT for large&#8209;scale Wi&#8209;Fi deployments, even from different vendors.

> This repository contains the open&#8209;source daemon wimoved of the WiMoVE system. If you want to learn more about the architecture of the system, take a look at our [docs](https://wimove-oss.github.io/docs/)

Reviewer: ```suggestion
This repository contains the open&#8209;source daemon wimoved of the WiMoVE system. If you want to learn more about the architecture of the system, take a look at our [documentation](https://wimove-oss.github.io/docs/architecture/).
```

## Setup

> Our documentation for the setup is still pretty early stage with some rough edges. If something doesn't work the way you expect or you get stuck please let us know, so we can improve our guides and help you with your issue. A hosted version of the setup guide is available [here](https://wimove-oss.github.io/docs/setup)

Reviewer: ```suggestion
Our documentation for the setup is still pretty early stage with some rough edges. If something doesn't work the way you expect or you get stuck please let us know, so we can improve our guides and help you with your issue. A hosted version of the setup guide is available [here](https://wimove-oss.github.io/docs/setup).
```

##  State of this software

:warning: This software is still work in progress and subject to change. :warning:

Currently we support OpenWRT with the following architectures: `ramips-mt7621`, `mvebu-cortexa9`. We test our software on the following Access Points: `ZyXEL NWA50AX` and `Linksys WRT1900ACS`. If you need support for another OpenWRT architecture feel free to open an issue. We'll consider adding another build pipeline. If you encounter an issue or have a feature request please open a ticket here on github. 

## Acknowledgements

> The article series of Vincent Bernat is a great starting point to learn about BGP EVPN with VXLAN. It helped us getting started so please check it out [Link](https://vincent.bernat.ch/en/blog/2017-vxlan-bgp-evpn)

Reviewer: ```suggestion
The article series of Vincent Bernat is a great starting point to learn about BGP EVPN with VXLAN. It helped us get started so please check it out [here](https://vincent.bernat.ch/en/blog/2017-vxlan-bgp-evpn).
```

![Linter Status](https://github.com/WiMoVE-OSS/wimoved/actions/workflows/linter.yml/badge.svg?branch=main)

WiMoVE is a scalable Wi-Fi System that partitions all stations in separate overlay L2 domains, to limit the amount of wireless L2 broadcast traffic. In large Wi-Fi systems broadcast traffic can take up large amounts of airtime. A great talk, that explains the problem in more detail is available [here](https://www.youtube.com/watch?v=v8y-r9JBhmw). The overlay L2 domains "follow" the stations, being resized on demand, thus preserving handover.

WiMoVE is built with standard network protocols, on top of open&#8209;source technology:

- The overlay networks use [BGP EVPN](https://www.rfc-editor.org/rfc/rfc8365.html) with [VXLAN](https://www.rfc-editor.org/rfc/rfc7348) encapsulation.
- All BGP speakers run [FRRouting](https://frrouting.org/) to exchange EVPN routes.
- The access points run [OpenWRT](https://openwrt.org/) with our custom, open&#8209;source daemon called [wimoved](https://github.com/wimove-oss/wimoved).

This solution allows for using commodity access points running OpenWRT for large&#8209;scale Wi&#8209;Fi deployments, even from different vendors.

> This repository contains the open&#8209;source daemon wimoved of the WiMoVE system. The daemon is responsible to handle events of hostapd and creating the VXLAN interfaces. The WiMoVE system consists of multiple other systems apart from the daemon. If you want to learn more about the architecture of the system or want to deploy your own setup, take a look at our [documentation](https://wimove-oss.github.io/docs/).

Reviewer 2: Language nitpick: _The daemon is reposible for handling hostapd events and creating VXLAN interfaces_

Reviewer 2: Instead of _the WiMoVE system_, I suggest putting _WiMoVE_

Reviewer 2: We should make clear that `wimoved` is supposed to run on the APs

Reviewer 2: Another language nitpick: _multiple other parts_ sounds better than _multiple other systems_ and is probably a little more accurate

##  Supported architectures

Currently we support OpenWRT with the following architectures: `ramips-mt7621`, `mvebu-cortexa9`. We test our software on the following Access Points: `ZyXEL NWA50AX` and `Linksys WRT1900ACS`. If you need support for another OpenWRT architecture feel free to open an issue.

## Development setup

> The development setup is an easy way to test out the WiMoVE daemon. Please be aware that after following this guide you do **not** have a working instance of the whole WiMoVE system. You can connect to the Wi-Fi but won't have any connectivity to other devices or the internet. Currently we only support Linux as development platform.

Reviewer 2: _won't have any connectivity_ -> _won't be connected_

    wpa_passphrase=<Secret key>
    per_sta_vif=1
    vlan_file=/etc/hostapd/hostapd.vlan
```

- Create a file `/etc/hostapd/hostapd.vlan` with the following content: 

```text
*   vlan#
```
- Hostapd has to be started with the option `-g /var/run/hostapd/global`. For the service to work accordingly you might have to edit the service file for hostapd. Run `systemctl status hostapd` to locate this file.
- Start `hostapd`, either with `systemctl start hostapd` or on the command line. You might need to stop `NetworkManager` before starting `hostapd` since the programs interfere with each other.

### Build WiMoVE

- Install [`libnl`](https://github.com/thom311/libnl). On a recent Linux system, the corresponding package is probably already installed.
- Install [`prometheus-cpp`](https://github.com/jupp0r/prometheus-cpp).
    - On Ubuntu, install the package `prometheus-cpp-dev`.
    - On Arch Linux, the package is available in the AUR as `prometheus-cpp-git`.
- Clone the repository by running `git clone https://github.com/WiMoVE-OSS/wimoved`.
- Build the project by running `cmake .` followed by `make -j$(nproc)`.
- Create a configuration file `wimoved.conf`

```text

> hapd_sock=/var/run/hostapd/<interface>

Reviewer 2: This is outdated, isn't it?

> ```text
> hapd_sock=/var/run/hostapd/<interface>
> ```

Reviewer: ```suggestion
```text
hapd_sockdir=/var/run/hostapd/
```
```

This solution allows for using commodity access points running OpenWRT for large&#8209;scale Wi&#8209;Fi deployments, even from different vendors.

## This repository

This repository contains the WiMoVE Access Point daemon wimoved.
The daemon is responsible for handling hostapd events and creating VXLAN interfaces.
WiMoVE consists of multiple other parts.
If you want to learn more about the architecture or create a full setup, take a look at our [documentation](https://wimove-oss.github.io/docs/).

## Supported architectures

Currently, we support OpenWRT with the architectures `ramips-mt7621` and `mvebu-cortexa9`.
We test our software on the Access Point models `ZyXEL NWA50AX` and `Linksys WRT1900ACS`.
If you need support for other OpenWRT architectures, feel free to open an issue.

## Development setup

The development setup is an easy way to test wimoved.
Please be aware that after following this guide, you do **not** have a full WiMoVE installation.
You can connect to the Wi-Fi but won't be connected to other devices or the Internet.

> Currently, we only support Linux as a development platform.

Author: Sure I will add that link.

- Start hostapd, either with `systemctl start hostapd` or on the command line. You might need to stop `NetworkManager` before starting hostapd since the programs interfere with each other.

### Build WiMoVE

- Install [`libnl`](https://github.com/thom311/libnl). On a recent Linux system, the corresponding package is probably already installed.
- Install [`prometheus-cpp`](https://github.com/jupp0r/prometheus-cpp).
    - On Ubuntu, install the package `prometheus-cpp-dev`.
    - On Arch Linux, the package is available in the AUR as `prometheus-cpp-git`.
- Clone the repository by running `git clone https://github.com/WiMoVE-OSS/wimoved`.
- Build the project by running `cmake .` followed by `make -j$(nproc)`.
- Start wimoved by running `./wimoved`.

### Coding guidelines

Format the source files by running `make format`.
Lint the source files by running `make lint`.
Build the tests by running `make test`.
Run both checks and the tests by doing `make precommit`.
The coding guidelines are enforced via the CI pipeline.

> As linting takes a long time, we recommend integrating clang-tidy into your editor.

Reviewer 3: I would prefer 
```suggestion
As linting takes a long time, we recommend integrating `clang-tidy` into your editor.
```
